<a href="https://colab.research.google.com/github/aissahm/Unicamp-MO431/blob/master/3o_Exerc%C3%ADcio_de_casa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

RA - 265189 - Aissa Hadj Mohamed

# **3o Exercício de casa**

We import the libraries needed for the project:

In [0]:
from scipy.optimize import minimize, rosen, rosen_der, line_search
import numpy as np

We install the following library to collect the execution time for each method. Also, as a side note, instead of printing the results of each minimize call, we decided to print them in a different cell as it takes a significative amount of time to just print the results.

In [16]:
!pip install ipython-autotime

%load_ext autotime

  Created wheel for ipython-autotime: filename=ipython_autotime-0.1-cp36-none-any.whl size=1832 sha256=88a4e9a1dedd8429eaca6d95c9046050530dbab5e71283341b2b0b07f7e45d77
  Stored in directory: /root/.cache/pip/wheels/d2/df/81/2db1e54bc91002cec40334629bc39cfa86dff540b304ebcd6e
Successfully built ipython-autotime


**We define below the Himmelblau function in two dimensions, as well as its gradient.**

In [0]:
def Himmelblau(x):
  return ( (x[0]**2 + x[1] - 11)**2 + ( x[0] + (x[1])**2 - 7 )**2 )

def Himmelblau_Jacobian(x):
  partial_x0 = 4 * x[0] * ( x[0]**2 + x[1] - 11 ) + 2*( x[0] + x[1]**2 - 7 )
  partial_x1 = 2*( x[0]**2 + x[1] - 11 )+ 4 * x[1] * ( x[0] + x[1]**2 - 7 ) 
  return np.array([partial_x0, partial_x1])

# **Conjugate gradient method**

Conjugate gradient method with the initial point at (4,4)

In [41]:
results = minimize(Himmelblau, [4, 4], method="CG")

time: 7.26 ms


In [42]:
print(results)

     fun: 1.9486292751344026e-15
     jac: array([ 5.80747415e-08, -1.19842767e-08])
 message: 'Optimization terminated successfully.'
    nfev: 68
     nit: 8
    njev: 17
  status: 0
 success: True
       x: array([2.99999999, 2.        ])
time: 1.64 ms


If we include the Jacobian of the Himmelblau function and pass it to the minimizer function:

In [44]:
results = minimize(Himmelblau, [4, 4], method="CG", jac=Himmelblau_Jacobian)

time: 3.18 ms


In [45]:
print(results)

     fun: 8.491206913057949e-21
     jac: array([-1.00791908e-09, -5.77447423e-10])
 message: 'Optimization terminated successfully.'
    nfev: 17
     nit: 8
    njev: 17
  status: 0
 success: True
       x: array([3., 2.])
time: 1.72 ms


When providing the Jacobian using the conjugate gradient method, the minimize function converges to the local minimum (3,2) as when we don't provide it, it needed the same amount of iterations (8) and the same number of evaluations for the Jacobian (17). However, it made lesser Himmelblau function evaluations ( 17 vs 68). Also, we can note that evaluation of the function at the local minimum given by the two methods is relatively different (8.491206913057949e-21 versus 1.9486292751344026e-15). This difference can be explained by the fact that the minimize function reaches the local minimum (2.99999999, 2.), without providing the Jacobian, and (3., 2.) when we do povide it. Also, we provided the Jacobian as an analytical expression which is more precise than computing it by numerical methods, when we don't pass it as a parameter inside the minimize function.  

# **BFGS**

BFGS com ponto inicial em (4,4)

In [52]:
results = minimize(Himmelblau, [4, 4], method="L-BFGS-B", jac=Himmelblau_Jacobian)

time: 2.2 ms


In [48]:
print(results)

      fun: 8.287611056879445e-13
 hess_inv: <2x2 LbfgsInvHessProduct with dtype=float64>
      jac: array([-6.76429348e-06,  3.62301182e-06])
  message: b'CONVERGENCE: NORM_OF_PROJECTED_GRADIENT_<=_PGTOL'
     nfev: 10
      nit: 9
   status: 0
  success: True
        x: array([2.99999986, 2.00000019])
time: 2.45 ms


In [54]:
results = minimize(Himmelblau, [4, 4], method="L-BFGS-B")

time: 2.29 ms


In [55]:
print(results)

      fun: 8.502778926721376e-13
 hess_inv: <2x2 LbfgsInvHessProduct with dtype=float64>
      jac: array([-6.76443727e-06,  3.62305493e-06])
  message: b'CONVERGENCE: NORM_OF_PROJECTED_GRADIENT_<=_PGTOL'
     nfev: 30
      nit: 9
   status: 0
  success: True
        x: array([2.99999985, 2.00000019])
time: 1.41 ms


As with the previous method (Conjugate of the Gradient),when providing the Jacobian using the gradient conjugate, the minimize function converges to the local minimum (2.99999985, 2.00000019) as when we don't provide it (2.99999986, 2.00000019), it needed the same amount of iterations (9). However, it made lesser Himmelblau function and gradient evaluations ( 10 vs 30). Also, we can note that evaluation of the function at the local minimum given by the two methods is different (8.287611056879445e-13 versus 8.502778926721376e-13). This difference can be explained by the fact that the minimize function reaches slighlty different values for the local minimum, with and without providing the Jacobian. Also, we provided the Jacobian as an analytical expression which is more precise than evaluating the gradient each time using numerical methods. 

# **Nelder-Mead**

Nelder-Mead with initial triangle with the following coordinates (-4,-4) (-4,1),(4,-1). We can note that even though below, we provide the initial point at (4,4), it won't be used by the minimize function as we provided the intial triangle as a parameter.

In [56]:
x0 = [[-4,-4], [-4,1],[4, -1]]

results = minimize(Himmelblau, [4,4], method="Nelder-Mead", options={'initial_simplex': x0})

time: 5.3 ms


In [57]:
print(results)

 final_simplex: (array([[ 3.58441449, -1.84811588],
       [ 3.58442434, -1.84817563],
       [ 3.5844613 , -1.84816348]]), array([1.06865670e-08, 3.75594501e-08, 6.84894824e-08]))
           fun: 1.0686566996168641e-08
       message: 'Optimization terminated successfully.'
          nfev: 77
           nit: 40
        status: 0
       success: True
             x: array([ 3.58441449, -1.84811588])
time: 2 ms


The local minimum found by the Nelder-Mead method is (3.58441449, -1.84811588). The value of the Himmelblau at this point is 1.0686566996168641e-08. The method needed 77 evaluations of the Himmelblau function and its gradient. 

# **NEWOA**

We install the library Py-BOBYQA for the NEWOA method.

In [0]:
pip install Py-BOBYQA

In [27]:
import pybobyqa

time: 918 µs


In [62]:
results = pybobyqa.solve(Himmelblau, [4, 4])

time: 123 ms


In [63]:
print(results)

****** Py-BOBYQA Results ******
Solution xmin = [3. 2.]
Objective value f(xmin) = 1.287703555e-21
Needed 58 objective evaluations (at 58 points)
Approximate gradient = [6.40524882e-09 5.63832257e-08]
Approximate Hessian = [[73.79866449 20.17722278]
 [20.17722278 34.16328261]]
Exit flag = 0
Success: rho has reached rhoend
******************************

time: 2.21 ms


The local minimum found by the method is (3, 2). The value of the Himmelblau at this point is 1.287703555e-21. The method needed 58 evaluations of the Himmelblau function as well as its gradient. 

# **Line search**

For the line search method, we start with the initial position of (4, 4). Then, we compute the gradient of the Himmelblau function at each point, and we take its negative value as a direction for the Line Search method. Below, we implement the Line Search method provided by the Scipy library and we iterate until we found a minimum for the Himmelblau function.

In [64]:
nmax = 100
x0 = [4,4]
xnew = x0
n = 0
function_evaluation_number = 0
gradient_evaluation_number = 0
prev_results = []

while True: 
  results = line_search(Himmelblau, Himmelblau_Jacobian, xnew, -Himmelblau_Jacobian(xnew)   )
  n = n + 1
  function_evaluation_number = function_evaluation_number + results[1]
  gradient_evaluation_number = gradient_evaluation_number + results[2]
  if n > nmax:
    break
  if results[0] == None:
    break
  prev_results = results
  xnew = xnew + results[0] * -Himmelblau_Jacobian(xnew)

time: 20.9 ms


/usr/local/lib/python3.6/dist-packages/scipy/optimize/linesearch.py:314: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)


In [65]:
print("Last iteration results =", prev_results)
print("Local minimum found = ", xnew)
print("Himmelblau value at the local minimum =", Himmelblau(xnew))
print("Number of iterations =", n)
print("Number of function evaluations =", function_evaluation_number)
print("Number of gradient evaluations =", gradient_evaluation_number)

Last iteration results = (0.008469284914264056, 7, 1, 7.888609052210118e-31, 9.150786500563737e-29, array([-1.77635684e-15,  1.16642198e-14]))
Local minimum found =  [-3.77931025 -3.28318599]
Himmelblau value at the local minimum = 7.888609052210118e-31
Number of iterations = 30
Number of function evaluations = 208
Number of gradient evaluations = 29
time: 5.31 ms


# **Conclusion**

In this exercise, we used 5 different methods, starting from the initial point (4, 4), in order to compute a local minimum of the Himmelblau function. It should be noted that this function contains 4 local minimum. 

Three methods (Conjugate of the Gradient, L-BFGS-B and NEWOA) found a local minimum at the point (3, 2). The most efficent one was L-BFGS-B, providing the Jacobian to the minimize function, with 9 iterations and only 2.2ms required for execution time.

The Nelder-Mead method found a local minimum at point (3.58441449, -1.84811588). It also took a relatively low time to execute, requiring just 5.3ms.

The Line Search method found a local minimum at the point (-3.77931025 -3.28318599). It made by far the highest number of Himmelblau function evaluations (195), but required only 20.9ms to execute.

The NEWOA optimization algorithm was by far the less efficient method, it required 123ms even though it needed a relatively average number of evaluations of the Himmelblau function of 58.
